rfcs/0010-eaglesong/0010-eaglesong.md

```diff
@@ -0,0 +1,335 @@
+
+The matrix is given below. A multiplication procedure follows.
+
+```
+bit_matrix = [
+    [1 1 1 1 0 1 0 1 1 1 1 1 0 0 0 1]
+    [0 1 1 1 1 0 1 0 1 1 1 1 1 0 0 1]
+    [0 0 1 1 1 1 0 1 0 1 1 1 1 1 0 1]
+    [0 0 0 1 1 1 1 0 1 0 1 1 1 1 1 1]
+    [1 1 1 1 1 0 1 0 1 0 1 0 1 1 1 0]
+    [1 0 0 0 1 0 0 0 1 0 1 0 0 1 1 1]
+    [1 0 1 1 0 0 0 1 1 0 1 0 0 0 1 0]
+    [1 0 1 0 1 1 0 1 0 0 1 0 0 0 0 1]
+    [0 1 0 1 0 1 1 0 1 0 0 1 0 0 0 1]
+    [0 0 1 0 1 0 1 1 0 1 0 0 1 0 0 1]
+    [0 0 0 1 0 1 0 1 1 0 1 0 0 1 0 1]
+    [0 0 0 0 1 0 1 0 1 1 0 1 0 0 1 1]
+    [1 1 1 1 0 0 0 0 1 0 0 1 1 0 0 0]
+    [0 1 1 1 1 0 0 0 0 1 0 0 1 1 0 0]
+    [0 0 1 1 1 1 0 0 0 0 1 0 0 1 1 0]
+    [1 1 1 0 1 0 1 1 1 1 1 0 0 0 1 1]
+]
+
+function ApplyBitMatrix( state_vector ):
+    for j in range[0:16] do:
+        output_vector[j] = 0
+        for k in range[0:16] do:
+            output_vector[j] = output_vector[j] xor (bit_matrix[j,k] * state_vector[k])
+    return output_vector
+```
+
+Note that the matrix multiplication needs involve only ARX operations. Since the coefficients of the matrix are known, the loops can be unrolled and the entire matrix multiplication can be described in terms of the xor operation alone. The use of the multiplication operator `*` is merely a notational convenience to stress the link with matrix multiplication.
+
+<a name="Circulant-Multiplication"></a>
+## Circulant Multiplication
+
+Every element of the state vector undergoes a map involving xor and rotation. Specifically, `state_vector[i] = state_vector[i] xor (state_vector[i] <<< a[i]) xor (state_vector[i] <<< b[i])`. The coefficients `a` and `b` are different for every component. They are given below, with a procedure for computing the entire step.
+
+```
+coefficients = [
+    [0, 2, 4],
+    [0, 13, 22],
+    [0, 4, 19],
+    [0, 3, 14],
+    [0, 27, 31],
+    [0, 3, 8],
+    [0, 17, 26],
+    [0, 3, 12],
+    [0, 18, 22],
+    [0, 12, 18],
+    [0, 4, 7],
+    [0, 4, 31],
+    [0, 12, 27],
+    [0, 7, 17],
+    [0, 7, 8],
+    [0, 1, 13]
+]
+
+function ApplyCirculantMultiplication( state_vector ):
+    for i in [0:16] do:
+        output_vector[i] = 0
+        for k in [0:3] do:
+            output_vector[i] = output_vector[i] xor (state_vector[i] <<< coefficients[i][k])
+    return output_vector
+```
+
+<a name="Injection-of-Constants"></a>
+## Injection of Constants
+
+A given constant is xored into every component of the state vector. The coefficients differ per component and per round. They are determined from using Keccak to expand the ASCII string 
+
+```
+The various ways in which the knowledge on which people base their plan is communicated to them is the crucial problem for any theory explaining the economic process, and the problem of what is the best way to utilizing knowledge initially dispersed among all the people is at least one of the main problems of economic policy - or of designing an efficient economic system.
```

Comment:
What's these words means?
It looks like nothing to do with the context.

Reply:
This quote by the Nobel prize winning Austrian economist FA Hayek captures concisely the Austrian school's perspective on the function of money within society: money and prices is the mechanism by which scarce resources are directed towards those needs most urgently wanted by society. Money, whether provided by a free market or an intervening state, lies at the heart of any sound understanding of society and how it operates -- not because everyone wants it but because of the problem it solves, i.e., the most-wanted allocation of scarce resources. Any theory of economics that fails to characterize the price mechanism as a network for rapid information dispersal, or free market money as a foundation for it, is necessarily incomplete.

This quote is relevant in the broader context of Nervos CKB because it describes exactly the problem landscape it was designed for. On the one hand, Nervos CKB offers a native currency, which enables a price mechanism and can in principle solve the distribution problem outright -- though more accurately it represents one building block among many with various engineering tradeoffs hoping to solve the same problem. On the other hand, the Common Knowledge Base (which is what CKB stands for) provides another type of knowledge, different from the one Hayek describes. In Hayek's perspective, the knowledge that makes society work is distributed across the minds of all individuals involved. In the CKB, certain knowledge is held and agreed upon by all participants. The interplay between Hayekian knowledge and consensus knowledge is the technical key ingredient that makes off-chain scaling solutions such as the Lightning Network work. By optimizing for the distribution of multi-functional; consensus knowledge, and by enabling the distribution of Hayekian knowledge as a result of a free price system, Nervos aims to provide new and better solutions to the economical problems that society faces.

In the specific context of Eaglesong, the purpose of the quote is to make the argument that the constants are indeed random and not chosen specifically to hide a trapdoor that only the designers know. The argument is 90% complete merely by using Keccak: the output is indistinguishable from random and yet uniquely determined by the input. The argument is completed for the remaining 10% by choosing an arbitrary fixed input string that simultaneously makes it obvious that it was not chosen because the other options lead to less usable output constants. For this purpose, what better arbitrary fixed input string is there to seed Keccak with, but the FA Hayek quote that captures Nervos CKB's target application landscape?